### ISSUE TITLE
Warning when using with webpack

### ISSUE THREAD

**Issue by fantapop** [lukechilds/keyv#45]
I'm importing [https://github.com/sindresorhus/got/](url) into my project and using webpack to build.   Webpack doesn't like requiring of expressions:

```
WARNING in ./node_modules/keyv/src/index.js
18:14-40 Critical dependency: the request of a dependency is an expression
 @ ./node_modules/keyv/src/index.js
 @ ./node_modules/cacheable-request/src/index.js
 @ ./node_modules/got/index.js
 @ ./lib/util.js
 @ ./functions/live-stream.js
```

I'm not sure what the best way to fix this up with would be

**Comment by lukechilds** [comment]
Does it work?

**Comment by fantapop** [comment]
It still seems to work. I haven't verified the code path it takes. I can
look into that. It makes the warning during build. It looked like the
pluggable  back end may have been opt in so I was assuming there was a
default.

**Comment by lukechilds** [comment]
It should work absolutely fine as long as your not including native Node.js storage adapters.

The Webpack warning is exactly that a warning, it's letting you know that one of your dependencies is requiring a dependency via an expression.

This is generally bad practice and will probably break static analysis but it's the only way to transparently require storage adapters.

It should be totally safe to ignore for now.

**Comment by roblav96** [comment]
@fantapop @lukechilds What's the recommended quick fix to remove the warnings from my console? I run a tight ship 📦:P

**Comment by lukechilds** [comment]
Close your eyes while the warning's on the screen.

**Comment by sindresorhus** [comment]
This has just another bad Webpack UX. I would recommend complaining on their issue tracker: https://github.com/webpack/webpack/issues/196

**Comment by fantapop** [comment]
I've been using the close my eyes method...

**Comment by Tobbe** [comment]
Don't know if this is good or bad, but I added this to my webpack config to get rid of the warning
```js
plugins: [
  new webpack.ContextReplacementPlugin(/keyv/),
],
```

**Comment by transitive-bullshit** [comment]
Does anyone know of a workaround to quiet the output for Next.js? (using next v12 with webpack v5)

**Comment by transitive-bullshit** [comment]
It turns out this isn't a wholly innocuous warning.

If you're using Next.js on Vercel, as part of their build optimizations for serverless functions (like dynamic page rendering), they trace module imports to remove as much unused code as possible.

It took me awhile to realize that while I'm using `keyv` with next.js v12 (and webpack v5) locally, everything would build and work fine, but when deployed to Vercel, any page renders or api calls would throw errors because next.js failed to find the expression-based `@keyv/redis` import.

If anyone else ends up here tracking this issue down, here was my fix: https://github.com/transitive-bullshit/nextjs-notion-starter-kit/commit/22262964498414324a9be59d0230f03e4df48527

While I can understand `keyv`'s design choice to have a simpler, implicit API, and while you could certainly argue that Next.js's behavior is aggressive, it's honestly a pretty common use case these days — and one that I think will only become more common in the future.

**Comment by transitive-bullshit** [comment]
https://github.com/microlinkhq/keyv by @Kikobeats appears to fix this issue, but upstream dependents like https://github.com/lukechilds/cacheable-request (which is used by `got`) would need to change. Or this package would need to update its API in the same way that `@keyvhq/core` has.

Either way, I think this is worth fixing as it is definitely affecting anyone using `got` with Next.js.

**Comment by Kikobeats** [comment]
@transitive-bullshit 

Although `@keyvhq/core` fix the thing, there are still place where the dependency will be not used (like `cacheable-request` sub-dependencies tree).

For these situations, you can declare a NPM override at your `package.json` 🙂

```
"overrides": {
    "cacheable-request": {
      "keyv": "npm:@keyvhq/core@~1.6.6"
    }
  }
```

**Comment by nickluger** [comment]
I use this package transitively via got and see this issue only, since got upgraded indirectly to `4.5.0` of `keyv`. 🤷‍♂️

**Comment by AuthorOfTheSurf** [comment]
In my case I needed to update my badly-outdated `@auth0/nextjs-auth0` dependency to the latest version. Some of my imports changed to look like `@auth0/nextjs-auth0/client` since they are client-side

My old auth0 dep was relying on `cacheable-request` and therefore `keyv` which I hadn't heard of till today. All of this can be seen in the error stack depending on your case

I encountered this bug when porting a NextJS 12 app to NextJS 13, so I suspect others may find their way here. In my case the error had a natural solution, and I was able to update an important security dependency
